src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs

```diff
@@ -178,5 +151,85 @@ public bool ManualSettingsOnly
 
             return useProxy;
         }
+
+        private void DetectScriptLocation()
+        {
+            WinHttpTraceHelper.Trace("WinInetProxyHelper.DetectScriptLocation: Start auto discovery.");
+            IntPtr autoProxyUrl = new IntPtr();
+
+            try
+            {
+                bool success = Interop.WinHttp.WinHttpDetectAutoProxyConfigUrl(
+                    Interop.WinHttp.WINHTTP_AUTO_DETECT_TYPE_DHCP | Interop.WinHttp.WINHTTP_AUTO_DETECT_TYPE_DNS_A,
+                    out autoProxyUrl);
+
+                if (success)
+                {
+                    // AutoDetect has precedence over the other settings.
+                    // If there is already a automatic configuration script location, override the value.
+                    AutoConfigUrl = Marshal.PtrToStringUni(autoProxyUrl);
+                    _useProxy = true;
+                }
+                else
+                {
+                    if (AutoConfigUrl == null)
+                    {
+                        WinHttpTraceHelper.Trace("WinInetProxyHelper.DetectScriptLocation: Auto discovery failed.");
+
+                        // This is to improve performance. For example, on home networks, where auto-detect will always
+                        // fail, but IE settings turn auto-detect ON by default. So in home networks on each
+                        // call we would try to retrieve the PAC location.
+
+                        // The downside of this approach is that if after some time the PAC file can be downloaded,
+                        // we will skip it.
+                        AutoDetect = false;
```

Comment:
As the email discussion with WinHttp team suggest, "The answer is if you want to be correct, you have to call the API (`WinHttpGetProxyForUrl`). Without re-implementing half of the proxy code in WinHttp and calling into tons of other components, you cannot know when it is Ok to skip it."

We will reach here when: 1. `WinHttpOpen` returns a invalid session handle, 2. `WinHttpGetIEProxyConfigForCurrentUser` not able to get automatic configuration script, 3. `WinHttpDetectAutoProxyConfigUrl` unable to auto discover the PAC URL (at this moment). It's highly likely that, there is no proxy used. But we are not 100% sure.

As the comment has suggested, there is a potential risk to set `AutoDetect = false` (which can cause `AutoSettingsUsed` returns false and skip calling into `WinHttpGetProxyForUrl`). On framework we decided to take the risk, and it already caused some issues there.

I'd like to hear people's feedback on the optimization in this case.

Reply:
>I'd like to hear people's feedback on the optimization in this case.

Can we quantify what the performance hit is if we always call the `WinHttpGetProxyForUrl` API for each request. It's hard to make tradeoff decisions if we don't understand what we're comparing exactly.

WinHttpGetProxyForUrl() already does internal caching of the PAC file whether it was found using WPAD protocol or explicitly specified. So, we aren't repeatedly doing WPAD protocol to find a PAC file for every request. So, I don't understand what the performance hit is in the steady-state case by calling this API. 

In cases where WinHttpGetIEProxyConfigForCurrentUser() has returned settings where "Automatically detect settings" or "Use automatic configuration script" are unchecked, it means that WPAD protcol shouldn't be used, and there is no explicit PAC file set.  So, that seems like the only case where we can avoid calling `WinHttpGetProxyForUrl` for each request.  The only other useful thing about WinHttpGetProxyForUrl would be to parse the proxy bypass list as specified in IE settings.  But, we decided to manually parse that list using managed code now.

We only call WinHttpGetIEProxyConfigForCurrentUser() once when the handler is created. That means that if the IE settings change while an application is still running, we won't be able to react to that.  In that case, only HttpClient/handler recycling can solve that.

Bottom line, I think trying to use WinHttpDetectAutoProxyConfigUrl() as a way to avoid calling WinHttpGetProxyForUrl() is not the safest approach in this PR.  I would prefer to use the algorithm I mentioned above regarding IE settings to determine whether or not we can skip calling WinHttpGetProxyForUrl().

Reply:
> I think trying to use WinHttpDetectAutoProxyConfigUrl() as a way to avoid calling WinHttpGetProxyForUrl() is not the safest approach in this PR.

I agree. I don't feel comfortable doing this as well, mainly because we don't know for sure when is ok to skip the `WinHttpGetProxyForUrl` call, which may result in some hard bugs in the future.

I will take the safe way and depend on IE settings. Will close this PR and submit a new one.

Reply:
>We need to make it as pay-for-play as we can.

I think if we want it pay-for-play, then the new PR #28965 covers that.  If a server wants optimum performance and avoid calling WinHttpGetProxyForUrl(), then it will make sure that the IE settings registry keys (if present) for "Use automatic configuration script" and "Automatically detect settings" are unchecked.